// angular-async-local-storage/angular-async-local-storage.js
import { Injectable, NgModule, PLATFORM_ID } from '@angular/core';
import { ReplaySubject } from 'rxjs/ReplaySubject';
import { first, map, mergeMap } from 'rxjs/operators';
import { fromEvent } from 'rxjs/observable/fromEvent';
import { of } from 'rxjs/observable/of';
import { _throw } from 'rxjs/observable/throw';
import { race } from 'rxjs/observable/race';
import { isPlatformBrowser } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @abstract
 */
class AsyncLocalDatabase {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class IndexedDBDatabase extends AsyncLocalDatabase {
    /**
     * Connects to IndexedDB
     */
    constructor() {
        super();
        /**
         * IndexedDB database name for local storage
         */
        this.dbName = 'ngStorage';
        /**
         * IndexedDB object store name for local storage
         */
        this.objectStoreName = 'localStorage';
        /**
         * IndexedDB key path name for local storage (where an item's key will be stored)
         */
        this.keyPath = 'key';
        /**
         * IndexedDB data path name for local storage (where items' value will be stored)
         */
        this.dataPath = 'value';
        /* Creating the RxJS ReplaySubject */
        this.database = new ReplaySubject();
        /* Connecting to IndexedDB */
        this.connect();
    }
    /**
     * Gets an item value in local storage
     * @template T
     * @param {?} key The item's key
     * @return {?} The item's value if the key exists, null otherwise, wrapped in an RxJS Observable
     */
    getItem(key) {
        /* Opening a trasaction and requesting the item in local storage */
        return this.transaction().pipe(map((transaction) => transaction.get(key)), mergeMap((request) => {
            /* Listening to the success event, and passing the item value if found, null otherwise */
            const /** @type {?} */ success = (/** @type {?} */ (fromEvent(request, 'success'))).pipe(map((event) => (/** @type {?} */ (event.target)).result), map((result) => result && (this.dataPath in result) ? (/** @type {?} */ (result[this.dataPath])) : null));
            /* Merging success and errors events and autoclosing the observable */
            return (/** @type {?} */ (race(success, this.toErrorObservable(request, `getter`))))
                .pipe(first());
        }));
    }
    /**
     * Sets an item in local storage
     * @param {?} key The item's key
     * @param {?} data The item's value, must NOT be null or undefined
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    setItem(key, data) {
        /* Storing null is not correctly supported by IndexedDB and unnecessary here */
        if (data == null) {
            return of(true);
        }
        /* Opening a transaction and checking if the item already exists in local storage */
        return this.getItem(key).pipe(map((existingData) => (existingData == null) ? 'add' : 'put'), mergeMap((method) => {
            /* Opening a transaction */
            return this.transaction('readwrite').pipe(mergeMap((transaction) => {
                let /** @type {?} */ request;
                /* Adding or updating local storage, based on previous checking */
                switch (method) {
                    case 'add':
                        request = transaction.add({ [this.dataPath]: data }, key);
                        break;
                    case 'put':
                    default:
                        request = transaction.put({ [this.dataPath]: data }, key);
                        break;
                }
                /* Merging success (passing true) and error events and autoclosing the observable */
                return (/** @type {?} */ (race(this.toSuccessObservable(request), this.toErrorObservable(request, `setter`))))
                    .pipe(first());
            }));
        }));
    }
    /**
     * Deletes an item in local storage
     * @param {?} key The item's key
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    removeItem(key) {
        /* Opening a transaction and checking if the item exists in local storage */
        return this.getItem(key).pipe(mergeMap((data) => {
            /* If the item exists in local storage */
            if (data != null) {
                /* Opening a transaction */
                return this.transaction('readwrite').pipe(mergeMap((transaction) => {
                    /* Deleting the item in local storage */
                    const /** @type {?} */ request = transaction.delete(key);
                    /* Merging success (passing true) and error events and autoclosing the observable */
                    return (/** @type {?} */ (race(this.toSuccessObservable(request), this.toErrorObservable(request, `remover`))))
                        .pipe(first());
                }));
            }
            /* Passing true if the item does not exist in local storage */
            return of(true).pipe(first());
        }));
    }
    /**
     * Deletes all items from local storage
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    clear() {
        /* Opening a transaction */
        return this.transaction('readwrite').pipe(mergeMap((transaction) => {
            /* Deleting all items from local storage */
            const /** @type {?} */ request = transaction.clear();
            /* Merging success (passing true) and error events and autoclosing the observable */
            return (/** @type {?} */ (race(this.toSuccessObservable(request), this.toErrorObservable(request, `clearer`))))
                .pipe(first());
        }));
    }
    /**
     * Connects to IndexedDB and creates the object store on first time
     * @return {?}
     */
    connect() {
        /* Connecting to IndexedDB */
        const /** @type {?} */ request = indexedDB.open(this.dbName);
        /* Listening the event fired on first connection, creating the object store for local storage */
        (/** @type {?} */ (fromEvent(request, 'upgradeneeded')))
            .pipe(first())
            .subscribe((event) => {
            /* Getting the database connection */
            const /** @type {?} */ database = /** @type {?} */ ((/** @type {?} */ (event.target)).result);
            /* Checking if the object store already exists, to avoid error */
            if (!database.objectStoreNames.contains(this.objectStoreName)) {
                /* Creating the object store for local storage */
                database.createObjectStore(this.objectStoreName);
            }
        });
        /* Listening the success event and converting to an RxJS Observable */
        const /** @type {?} */ success = /** @type {?} */ (fromEvent(request, 'success'));
        /* Merging success and errors events */
        (/** @type {?} */ (race(success, this.toErrorObservable(request, `connection`))))
            .pipe(first())
            .subscribe((event) => {
            /* Storing the database connection for further access */
            this.database.next(/** @type {?} */ ((/** @type {?} */ (event.target)).result));
        }, (error) => {
            this.database.error(/** @type {?} */ (error));
        });
    }
    /**
     * Opens an IndexedDB transaction and gets the local storage object store
     * @param {?=} mode Default to 'readonly' for read operations, or 'readwrite' for write operations
     * @return {?} An IndexedDB transaction object store, wrapped in an RxJS Observable
     */
    transaction(mode = 'readonly') {
        /* From the IndexedDB connection, opening a transaction and getting the local storage objet store */
        return this.database
            .pipe(map((database) => database.transaction([this.objectStoreName], mode).objectStore(this.objectStoreName)));
    }
    /**
     * Transforms a IndexedDB success event in an RxJS Observable
     * @param {?} request The request to listen
     * @return {?} A RxJS Observable with true value
     */
    toSuccessObservable(request) {
        /* Transforming a IndexedDB success event in an RxJS Observable with true value */
        return (/** @type {?} */ (fromEvent(request, 'success')))
            .pipe(map(() => true));
    }
    /**
     * Transforms a IndexedDB error event in an RxJS ErrorObservable
     * @param {?} request The request to listen
     * @param {?=} error Optionnal details about the error's origin
     * @return {?} A RxJS ErrorObservable
     */
    toErrorObservable(request, error = ``) {
        /* Transforming a IndexedDB error event in an RxJS ErrorObservable */
        return (/** @type {?} */ (fromEvent(request, 'error')))
            .pipe(mergeMap((event) => _throw(new Error(`IndexedDB ${error} issue : ${request.error.message}.`))));
    }
}
IndexedDBDatabase.decorators = [
    { type: Injectable },
];
/** @nocollapse */
IndexedDBDatabase.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class LocalStorageDatabase extends AsyncLocalDatabase {
    constructor() {
        super(...arguments);
        /* Initializing native localStorage right now to be able to check its support on class instanciation */
        this.localStorage = localStorage;
    }
    /**
     * Gets an item value in local storage
     * @template T
     * @param {?} key The item's key
     * @return {?} The item's value if the key exists, null otherwise, wrapped in an RxJS Observable
     */
    getItem(key) {
        const /** @type {?} */ unparsedData = this.localStorage.getItem(key);
        let /** @type {?} */ parsedData = null;
        if (unparsedData != null) {
            try {
                parsedData = JSON.parse(unparsedData);
            }
            catch (/** @type {?} */ error) {
                return _throw(new Error(`Invalid data in localStorage.`));
            }
        }
        return of(parsedData);
    }
    /**
     * Sets an item in local storage
     * @param {?} key The item's key
     * @param {?} data The item's value, must NOT be null or undefined
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    setItem(key, data) {
        this.localStorage.setItem(key, JSON.stringify(data));
        return of(true);
    }
    /**
     * Deletes an item in local storage
     * @param {?} key The item's key
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    removeItem(key) {
        this.localStorage.removeItem(key);
        return of(true);
    }
    /**
     * Deletes all items from local storage
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    clear() {
        this.localStorage.clear();
        return of(true);
    }
}
LocalStorageDatabase.decorators = [
    { type: Injectable },
];
/** @nocollapse */
LocalStorageDatabase.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MockLocalDatabase extends AsyncLocalDatabase {
    constructor() {
        super(...arguments);
        this.localStorage = new Map();
    }
    /**
     * Gets an item value in local storage
     * @template T
     * @param {?} key The item's key
     * @return {?} The item's value if the key exists, null otherwise, wrapped in an RxJS Observable
     */
    getItem(key) {
        const /** @type {?} */ rawData = this.localStorage.get(key);
        return of((rawData !== undefined) ? rawData : null);
    }
    /**
     * Sets an item in local storage
     * @param {?} key The item's key
     * @param {?} data The item's value, must NOT be null or undefined
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    setItem(key, data) {
        this.localStorage.set(key, data);
        return of(true);
    }
    /**
     * Deletes an item in local storage
     * @param {?} key The item's key
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    removeItem(key) {
        this.localStorage.delete(key);
        return of(true);
    }
    /**
     * Deletes all items from local storage
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    clear() {
        this.localStorage.clear();
        return of(true);
    }
}
MockLocalDatabase.decorators = [
    { type: Injectable },
];
/** @nocollapse */
MockLocalDatabase.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * \@todo Add other JSON Schema validation features
 */
class JSONValidator {
    constructor() {
        this.simpleTypes = ['string', 'number', 'boolean', 'object'];
    }
    /**
     * @param {?} value
     * @return {?}
     */
    isObjectNotNull(value) {
        return (value !== null) && (typeof value === 'object');
    }
    /**
     * Validate a JSON data against a JSON Schema
     * @param {?} data JSON data to validate
     * @param {?} schema Subset of JSON Schema
     * @return {?} If data is valid : true, if it is invalid : false, and throws if the schema is invalid
     */
    validate(data, schema) {
        if (!this.isObjectNotNull(schema)) {
            throw new Error(`A schema must be an object (unlike spec, booleans are not supported to enforce strict types).`);
        }
        if ((!schema.hasOwnProperty('type') || schema.type === 'array' || schema.type === 'object')
            && !schema.hasOwnProperty('properties') && !schema.hasOwnProperty('items')) {
            throw new Error(`Each value must have a 'type' or 'properties' or 'items', to enforce strict types.`);
        }
        if (schema.hasOwnProperty('type') && !this.validateType(data, schema)) {
            return false;
        }
        if (schema.hasOwnProperty('items') && !this.validateItems(data, schema)) {
            return false;
        }
        if (schema.hasOwnProperty('properties')) {
            if (schema.hasOwnProperty('required') && !this.validateRequired(data, schema)) {
                return false;
            }
            if (!this.validateProperties(data, schema)) {
                return false;
            }
        }
        return true;
    }
    /**
     * @param {?} data
     * @param {?} schema
     * @return {?}
     */
    validateProperties(data, schema) {
        if (!this.isObjectNotNull(data)) {
            return false;
        }
        if (!schema.properties || !this.isObjectNotNull(schema.properties)) {
            throw new Error(`'properties' must be a schema object.`);
        }
        /**
             * Check if the object doesn't have more properties than expected
             * Equivalent of additionalProperties: false
             */
        if (Object.keys(schema.properties).length !== Object.keys(data).length) {
            return false;
        }
        /* Recursively validate all properties */
        for (let /** @type {?} */ property in schema.properties) {
            if (schema.properties.hasOwnProperty(property) && data.hasOwnProperty(property)) {
                if (!this.validate(data[property], schema.properties[property])) {
                    return false;
                }
            }
        }
        return true;
    }
    /**
     * @param {?} data
     * @param {?} schema
     * @return {?}
     */
    validateRequired(data, schema) {
        if (!this.isObjectNotNull(data)) {
            return false;
        }
        if (!Array.isArray(schema.required)) {
            throw new Error(`'required' field must be an array. Note that since JSON Schema draft 6, booleans are not supported anymore.`);
        }
        for (let /** @type {?} */ requiredProp of schema.required) {
            if (typeof requiredProp !== 'string') {
                throw new Error(`'required' array must contain strings only.`);
            }
            /* Checks if the property is present in the schema 'properties' */
            if (!schema.properties || !schema.properties.hasOwnProperty(requiredProp)) {
                throw new Error(`'required' properties must be described in 'properties' too.`);
            }
            /* Checks if the property is present in the data */
            if (!data.hasOwnProperty(requiredProp)) {
                return false;
            }
        }
        return true;
    }
    /**
     * @param {?} data
     * @param {?} schema
     * @return {?}
     */
    validateType(data, schema) {
        if (Array.isArray(schema.type)) {
            return this.validateTypeList(data, schema);
        }
        if (typeof schema.type !== 'string') {
            throw new Error(`'type' must be a string (arrays of types are not supported yet).`);
        }
        if ((schema.type === 'null') && (data !== null)) {
            return false;
        }
        if ((this.simpleTypes.indexOf(schema.type) !== -1) && (typeof data !== schema.type)) {
            return false;
        }
        if ((schema.type === 'integer') && ((typeof data !== 'number') || !Number.isInteger(data))) {
            return false;
        }
        return true;
    }
    /**
     * @param {?} data
     * @param {?} schema
     * @return {?}
     */
    validateTypeList(data, schema) {
        const /** @type {?} */ types = /** @type {?} */ (schema.type);
        const /** @type {?} */ typesTests = [];
        for (let /** @type {?} */ type of types) {
            typesTests.push(this.validateType(data, { type }));
        }
        return (typesTests.indexOf(true) !== -1);
    }
    /**
     * @param {?} data
     * @param {?} schema
     * @return {?}
     */
    validateItems(data, schema) {
        if (!Array.isArray(data)) {
            return false;
        }
        if (Array.isArray(schema.items)) {
            return this.validateItemsList(data, schema);
        }
        if (!schema.items || !this.isObjectNotNull(schema.items)) {
            throw new Error(`'items' must be a schema object.`);
        }
        for (let /** @type {?} */ value of data) {
            if (!this.validate(value, schema.items)) {
                return false;
            }
        }
        return true;
    }
    /**
     * @param {?} data
     * @param {?} schema
     * @return {?}
     */
    validateItemsList(data, schema) {
        const /** @type {?} */ items = /** @type {?} */ (schema.items);
        if (data.length !== items.length) {
            return false;
        }
        for (let /** @type {?} */ i = 0; i < items.length; i += 1) {
            if (!this.validate(data[i], items[i])) {
                return false;
            }
        }
        return true;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */

class AsyncLocalStorage {
    /**
     * @param {?} database
     * @param {?} jsonValidator
     */
    constructor(database, jsonValidator) {
        this.database = database;
        this.jsonValidator = jsonValidator;
        this.getItemOptionsDefault = {
            schema: null
        };
    }
    /**
     * Gets an item value in local storage
     * @template T
     * @param {?} key The item's key
     * @param {?=} options
     * @return {?} The item's value if the key exists, null otherwise, wrapped in an RxJS Observable
     */
    getItem(key, options = this.getItemOptionsDefault) {
        return this.database.getItem(key).pipe(/* Validate data upon a json schema if requested */
        mergeMap((data) => {
            if (options.schema && data !== null) {
                let /** @type {?} */ validation = true;
                try {
                    validation = this.jsonValidator.validate(data, options.schema);
                }
                catch (/** @type {?} */ error) {
                    return _throw(error);
                }
                if (!validation) {
                    return _throw(new Error(`JSON invalid`));
                }
            }
            return of(data);
        }));
    }
    /**
     * Sets an item in local storage
     * @param {?} key The item's key
     * @param {?} data The item's value, must NOT be null or undefined
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    setItem(key, data) {
        return this.database.setItem(key, data);
    }
    /**
     * Deletes an item in local storage
     * @param {?} key The item's key
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    removeItem(key) {
        return this.database.removeItem(key);
    }
    /**
     * Deletes all items from local storage
     * @return {?} An RxJS Observable to wait the end of the operation
     */
    clear() {
        return this.database.clear();
    }
}
AsyncLocalStorage.decorators = [
    { type: Injectable },
];
/** @nocollapse */
AsyncLocalStorage.ctorParameters = () => [
    { type: AsyncLocalDatabase, },
    { type: JSONValidator, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @param {?} platformId
 * @return {?}
 */
function databaseFactory(platformId) {
    if (isPlatformBrowser(platformId) && ('indexedDB' in window)) {
        /* Try with IndexedDB in modern browsers */
        return new IndexedDBDatabase();
    }
    else if (isPlatformBrowser(platformId) && ('localStorage' in window)) {
        /* Try with localStorage in old browsers (IE9) */
        return new LocalStorageDatabase();
    }
    else {
        /* Fake database for server-side rendering (Universal) */
        return new MockLocalDatabase();
    }
}
class AsyncLocalStorageModule {
}
AsyncLocalStorageModule.decorators = [
    { type: NgModule, args: [{
                providers: [
                    JSONValidator,
                    {
                        provide: AsyncLocalDatabase,
                        useFactory: databaseFactory,
                        deps: [PLATFORM_ID]
                    },
                    AsyncLocalStorage,
                ]
            },] },
];
/** @nocollapse */
AsyncLocalStorageModule.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { AsyncLocalDatabase, IndexedDBDatabase, LocalStorageDatabase, MockLocalDatabase, JSONValidator, AsyncLocalStorage, AsyncLocalStorageModule, databaseFactory as ɵa };
